### PR TITLE
add trainer.stop and fix a bug for train_by_parallel_executor

### PR DESCRIPTION
**Pull request by jacquesqiao**
fix: https://github.com/PaddlePaddle/Paddle/issues/10754

### CODE REVIEW

File: python/paddle/fluid/tests/book/high-level-api/fit_a_line/test_fit_a_line.py [hunk 1]
```diff
@@ -57,22 +57,20 @@ def train(use_cuda, train_program, save_dirname):
...
-            
-            ...
-            ['25.768919467926025']
-            ['15.343549569447836']
-            ...
-            
-            '''
-            if float(test_metrics[0]) < 20.0:
+        if isinstance(event, fluid.EndStepEvent):
+            if event.step == 10:
```
Comment (sidgoyal78) [inline python/paddle/fluid/tests/book/high-level-api/fit_a_line/test_fit_a_line.py]: Is this step for epoch? Or is it more like for iterations (different mini-batches across epochs)?

File: python/paddle/fluid/trainer.py [hunk 2]
```diff
@@ -327,9 +336,7 @@ def _train_by_parallel_executor(self, num_epochs, event_handler, reader,
             feeder = data_feeder.DataFeeder(
                 feed_list=feed_var_list, place=self.place)
             reader = feeder.decorate_reader(reader, multi_devices=True)
-            for epoch_id in range(num_epochs):
-                self._train_by_any_executor(event_handler, pe, num_epochs,
-                                            reader)
+            self._train_by_any_executor(event_handler, pe, num_epochs, reader)
```
Comment (daming-lu) [inline python/paddle/fluid/trainer.py]: So what is the reason of getting rid of the `for` loop? Just curious 😁

Comment (daming-lu) [reply python/paddle/fluid/trainer.py]: 👍

File: python/paddle/fluid/tests/book/high-level-api/fit_a_line/test_fit_a_line.py (hunk 1 above)
Comment (sidgoyal78) [reply python/paddle/fluid/tests/book/high-level-api/fit_a_line/test_fit_a_line.py]: Yeah, actually I was just confused about the naming. I was thinking that `event.step` should represent iterations. Probably, iterations should never counter back to 0. But in this case, we have `event.step` based on step_id. 

However, the step_id is from 0 to number of mini-batches. And after each epoch the step_id is reset to 0. 

So basically the point was that could be have a separate naming convention :)

**Pull request merged**